File: python/binary_search/33_Search_in_Rotated_Sorted_Array.py
'''
Time: O(logn)
'''

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l ,r = 0, len(nums)-1

        while l <= r:  # =: for case like [x]
            mid = (l + r) // 2
            if nums[mid] == target:
                return mid

            # if mid is in left portion
            if nums[l] <= nums[mid]:  # if len(nums) == 2, l = 0, r = 1, mid = 0, mid == l
                if (target <= nums[mid]) and (target >= nums[l]):
                    r = mid - 1
                else:
                    l = mid + 1
            # if mid is in right portion
            else:
                if (target >= nums[mid]) and (target <= nums[r]):
                    l = mid + 1
                else:
                    r = mid - 1

        return -1

File: javascript/binary_search/33_Search_in_Rotated_Sorted_Array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let l = 0, r = nums.length - 1;

    while (l <= r) {
        mid = Math.floor((l + r) / 2);
        if (nums[mid] == target) return mid;

        // mid in left portion
        if (nums[l] <= nums[mid]) {
            if (target < nums[mid] && target >= nums[l]) r = mid - 1;
            else l = mid + 1;
        }
        // mid in right portion
        else {
            if (target > nums[mid] && target <= nums[r]) l = mid + 1;
            else r = mid - 1;
        }
    }
    return -1;
};